src/conn/conn_compact.c

```diff
@@ -0,0 +1,196 @@
+             */
+            __wt_sleep(10, 0);
+        }
+
+        /* Only an explicit call to the compact API should signal the thread to wake up. */
+        if (!conn->background_compact.running) {
+            /*
+             * TODO: Depending on the previous state, we may not want to clear out the last key
+             * used. This could be useful if the server was paused to be resumed later.
+             */
+            key = NULL;
+            __wt_cond_wait(session, conn->background_compact.cond, 0, NULL);
+        }
+
+        /*
+         * Now the server has been requested to run, check if it is supposed to, we could be
+         * quitting or reconfiguring the connection too.
+         */
+        if (!__compact_server_run_chk(session))
+            break;
+
+        /* Open a metadata cursor. */
+        WT_ERR(__wt_metadata_cursor(session, &cursor));
+
+        /*
+         * If we are in the middle of an iteration, position the cursor on the previously selected
+         * key.
+         */
+        if (key != NULL) {
+            cursor->set_key(cursor, key);
+            WT_ERR(cursor->search_near(cursor, &exact));
+        }
+
+        /*
+         * If no key has been set yet, move to the first record. If the previously used key could
+         * not be found, make sure not to go backwards.
+         */
+        if (key == NULL || exact <= 0)
+            WT_ERR_NOTFOUND_OK(cursor->next(cursor), true);
+
+        /* Find a table to compact. */
+        while (ret == 0) {
+            WT_ERR(cursor->get_key(cursor, &key));
+            if (WT_PREFIX_MATCH(key, "file:") && !WT_STREQ(key, WT_HS_URI))
+                /* FIXME-WT-11343: check if the table is supposed to be compacted. */
+                break;
+            WT_ERR_NOTFOUND_OK(cursor->next(cursor), true);
+        }
+
+        /* We may have reached the end of the file. */
+        if (ret == WT_NOTFOUND) {
+            WT_ERR(__wt_metadata_cursor_release(session, &cursor));
+            full_iteration = true;
+            key = NULL;
+            continue;
+        }
+
+        /* Always close the metadata cursor. */
+        WT_ERR(__wt_metadata_cursor_release(session, &cursor));
+
+        /* Compact the file. */
+        /* TODO: Make sure we are not reading a cached configuration. */
+        ret = wt_session->compact(wt_session, key, conn->background_compact.cfg);
+        /* FIXME-WT-11343: compaction is done, update the data structure for this table. */
+        /*
+         * Compact may return:
+         * - EBUSY for various reasons.
+         * - ETIMEDOUT if the timer has been configured and compaction took too long. TODO: Can the
+         * background compaction be configured with a timeout?
+         * - WT_NOTFOUND if the underlying file has been deleted.
+         */
+        if (ret == EBUSY || ret == WT_NOTFOUND || ret == ETIMEDOUT)
+            ret = 0;
+        WT_ERR(ret);
+    }
+
+    WT_ERR(__wt_metadata_cursor_close(session));
```

Comment:
We should move this to after the `err:` label

Reply:
In the `err` section, `__wt_panic` is called. Having a metadata cursor opened should not matter too much when WT panics. I am happy to have @keitharnoldsmith and @kommiharibabu's opinions though.

Reply:
I see, that makes sense.

Reply:
Yes, I agree that there is no need to close the metadata cursor as we are going to shutdown the server with PANIC failure.